stdlib/asm/math/poly512.masm

```diff
@@ -0,0 +1,6682 @@
+use.std::math::ntt512
+use.std::math::u64
+
+# Given two consecutive words on stack, this routine performs 
+# element wise multiplication, while keeping resulting single
+# word on stack.
+#
+# Expected stack state looks like
+#
+# [a0, a1, a2, a3, b0, b1, b2, b3]
+#
+# What this routine does is
+#
+# c`i` = a`i` * b`i` mod P | i ∈ [0, 4), P = 2 ^ 64 - 2 ^ 32 + 1
+#
+# Output stack state looks like
+#
+# [c0, c1, c2, c3]
+proc.mul_word
+    movup.4
+    mul
+
+    swap
+    movup.4
+    mul
+    swap
+
+    movup.2
+    movup.4
+    mul
+    movdn.2
+
+    movup.3
+    movup.4
+    mul
+    movdn.3
+end
+
+# Given two consecutive words on stack, this routine performs 
+# element wise addition, while keeping resulting single
+# word on stack.
+#
+# Expected stack state looks like
+#
+# [a0, a1, a2, a3, b0, b1, b2, b3]
+#
+# What this routine does is
+#
+# c`i` = a`i` + b`i` mod P | i ∈ [0, 4), P = 2 ^ 64 - 2 ^ 32 + 1
+#
+# Output stack state looks like
+#
+# [c0, c1, c2, c3]
+proc.add_word
+    movup.4
+    add
+
+    swap
+    movup.4
+    add
+    swap
+
+    movup.2
+    movup.4
+    add
+    movdn.2
+
+    movup.3
+    movup.4
+    add
+    movdn.3
+end
+
+# Given two operands ( i.e. field elements a, b ) on stack top, this routine computes c = a % b
+#
+# Expected stack state
+#
+# [b, a, ...]
+#
+# Output stack state looks like
+#
+# [c, ...] | c = a % b
+proc.mod_div
+    swap
+    u32split
+    movup.2
+    u32split
+    exec.u64::unchecked_mod
+    drop # higher 32 -bit must always be 0, because b = 12289
```

Comment:
`u64::unchecked_mod` is quite expensive. Since we know that `b=12289` I think we can implement a much more efficient version (potentially twice as efficient). Given this this function is used quite frequently, I think it may be worth it.

Reply:
You're correct @bobbinth . New implementation of modular reduction by 12289, decreases cost from 71 cycles to 39 cycles. See commit 279afec